metadata/src/main/java/org/apache/kafka/timeline/SnapshottableHashTable.java

```diff
@@ -0,0 +1,449 @@
+            return baseSize();
+        } else {
+            Iterator<Snapshot> iterator = snapshotRegistry.iterator(epoch);
+            while (iterator.hasNext()) {
+                Snapshot snapshot = iterator.next();
+                HashTier<T> tier = snapshot.getDelta(SnapshottableHashTable.this);
+                if (tier != null) {
+                    return tier.size;
+                }
+            }
+            return baseSize();
+        }
+    }
+
+    T snapshottableGet(Object key, long epoch) {
+        T result = baseGet(key);
+        if (result != null && result.startEpoch() <= epoch) {
+            return result;
+        }
+        if (epoch == LATEST_EPOCH) {
+            return null;
+        }
+        Iterator<Snapshot> iterator = snapshotRegistry.iterator(epoch);
+        while (iterator.hasNext()) {
+            Snapshot snapshot = iterator.next();
+            HashTier<T> tier = snapshot.getDelta(SnapshottableHashTable.this);
+            if (tier != null && tier.deltaTable != null) {
+                result = tier.deltaTable.baseGet(key);
+                if (result != null) {
+                    if (result.startEpoch() <= epoch) {
+                        return result;
+                    } else {
+                        return null;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    boolean snapshottableAddUnlessPresent(T object) {
+        T prev = baseGet(object);
+        if (prev != null) {
+            return false;
+        }
+        object.setStartEpoch(snapshotRegistry.latestEpoch() + 1);
+        int prevSize = baseSize();
+        baseAddOrReplace(object);
+        updateTierData(prevSize);
+        return true;
+    }
+
+    T snapshottableAddOrReplace(T object) {
+        object.setStartEpoch(snapshotRegistry.latestEpoch() + 1);
+        int prevSize = baseSize();
+        T prev = baseAddOrReplace(object);
+        if (prev == null) {
+            updateTierData(prevSize);
+        } else {
+            updateTierData(prev, prevSize);
+        }
+        return prev;
+    }
+
+    Object snapshottableRemove(Object object) {
+        T prev = baseRemove(object);
+        if (prev == null) {
+            return null;
+        } else {
+            updateTierData(prev, baseSize() + 1);
+            return prev;
+        }
+    }
+
+    private void updateTierData(int prevSize) {
+        Iterator<Snapshot> iterator = snapshotRegistry.reverseIterator();
+        if (iterator.hasNext()) {
+            Snapshot snapshot = iterator.next();
+            HashTier<T> tier = snapshot.getDelta(SnapshottableHashTable.this);
+            if (tier == null) {
+                tier = new HashTier<>(prevSize);
+                snapshot.setDelta(SnapshottableHashTable.this, tier);
+            }
+        }
+    }
+
+    private void updateTierData(T prev, int prevSize) {
+        Iterator<Snapshot> iterator = snapshotRegistry.reverseIterator();
```

Comment:
Perhaps we could have a reverseIterator taking a starting snapshot similar to the forward iterator.

Reply:
In this case we actually just want the very latest snapshot, so we don't need a starting position